nixos: doc: implement related packages in the manual (again)

oxij opened the pull request:
###### What is this?

#32424 reverted by #33006 brought back to you with laziness and `meta.evaluates` of #33057.
Closes #33007.

###### Things done

- [X] Seems to work.
- [X] `configuration.nix(5)` manual builds on x86_64-linux.
- [ ] `configuration.nix(5)` manual builds on aarch64-linux.

Please check the last item with OfBorg before merging, I'm too lazy to rebuild aarch64 stdenv just to test this.

---

/cc @grahamc @vcunat

oxij committed:
nixos/doc: push all the `enable*' and `package*` options to the top of their option group

Why? Because this way configuration.nix(5) can be read linearly.

Before:

> virtualisation.xen.bootParams
> ...
> virtualisation.xen.enable
> ...
> virtualisation.xen.package
> ...

After:

> virtualisation.xen.enable
> virtualisation.xen.package
> virtualisation.xen.bootParams
> ...

oxij committed:
nixos, lib: implement relatedPackages option

This allows one to specify "related packages" in NixOS that get rendered into
the configuration.nix(5) man page. The interface philosophy is pretty much
stolen from TeX bibliography.

See the next several commits for examples.

oxij committed:
lib: implement `compare`, `splitByAndCompare`, and `compareLists`

oxij committed:
nixos/tmux: add related package

This is a trivial example of `relatedPackages` option usage.

oxij committed:
xen, qemu: passthru the path to qemu-system-i386

oxij committed:
nixos/xen-dom0: add related packages, make it play well with them

This is a custom attribute set example of `relatedPackages` option usage.

oxij commented:
Well, it seems to work on my hacky cross-compiled aarch64 stdenv.

Ping @grahamc again because you was involved in 3 out of 4 related PRs.

grahamc commented on lib/default.nix:
```diff
@@ -71,8 +71,8 @@ let
-      reverseList listDfs toposort sort take drop sublist last init
-      crossLists unique intersectLists subtractLists
+      reverseList listDfs toposort sort compareLists take drop sublist
+      last init crossLists unique intersectLists subtractLists
```
One of the intentions behind making the inherits explicit here is so we don't need to make all the lib functions available in the top level, but instead say `lib.lists.compareLists`. If this seems like a reasonable thing to do, would you like to not `inherit` them here? It could be we're just not ready for that yet, seeing as I don't think many functions aren't exposed at the top level here.

grahamc commented on lib/trivial.nix:
```diff
@@ -81,6 +81,31 @@ rec {
+     compare elements of the same subtype with `yes` and `no`
+     comparisons respectively.
+  */
+  splitByAndCompare = p: yes: no: a: b:
```
Please provide a usage example for these functions.

grahamc replied:
Also, is this really trivial?

grahamc commented on nixos/doc/manual/default.nix:
```diff
@@ -32,8 +48,22 @@ let
+
   # Convert the list of options into an XML file.
-  optionsXML = builtins.toFile "options.xml" (builtins.toXML optionsList');
+  optionsXML = builtins.toFile "options.xml" (builtins.toXML optionsList'');
```
I'm worried about the `''` being confusing here, since it is identical to otherwise valid Nix syntax. Is there a better variable name you can imagine?

grahamc commented:
Will `relatedPackages` cover cases like `python3Packages.requests`? What will happen if a relatedPackage is deleted, and the reference is left dangling?

oxij commented:
> One of the intentions behind making the inherits explicit here is so we don't need to make all the lib functions available in the top level, but instead say `lib.lists.compareLists`. If this seems like a reasonable thing to do, would you like to not `inherit` them here? It could be we're just not ready for that yet, seeing as I don't think many functions aren't exposed at the top level here.

Hm. I interpret those things differently. I though the intention of doing it this way was that you should inherit anything generic and not inherit things that are unlikely to be used out of the lib. Those new functions are pretty generic as proved by their usage out of the lib. 

Are you sure you want to (eventually) soil every nix module with a bunch of `with lib.<something>` at the top (like C++ does with `use <something>`)?
Why isn't `with lib` enough? I would prefer `lib` to be like Haskell's `Prelude` that gives the cohesive set of things you most likely want by default.

> Will `relatedPackages` cover cases like `python3Packages.requests`?

Not as a string because I didn't want to deal with parsing attributes (and the way `rename.nix` does it is super-ugly).
You can specify `{ name = "python3Packages.requests"; package = pkgs.python3Packages.requests; }`, though.

However, its is a valid question whenever something better can be invented, like, say, allowing both simple strings and ugly lists of `rename.nix`. I'll think about it.

> What will happen if a relatedPackage is deleted, and the reference is left dangling?

An error. I think that's a good thing. You are not supposed to delete packages that are referenced in documentation.

> Please provide a usage example for these functions.

The code I introduced next does provide it, but, okay, will fix.

> Also, is this really trivial?

Well, I know some people fear formulas, but it is the best way I have to describe what it does without giving in-words description of the actual implementation. Feel free to propose a better one.

> I'm worried about the `''` being confusing here, since it is identical to otherwise valid Nix syntax. Is there a better variable name you can imagine? 

Will fix.

grahamc commented:
I agree with everything you said, sounds great! I look forward to the few remaining fixes :)

oxij committed:
nixos/adb: add related package

This is an attribute path example of `relatedPackages` option usage.

oxij commented:
ping? Just as I said, its completely finished now.
I added a some more documentation beyond what was discussed and implemented the `x.y.z` with `rename.nix`-like lists. Seems perfect to me.

grahamc commented:
![image](https://user-images.githubusercontent.com/76716/36057153-dbfe4178-0dd8-11e8-827f-48eaae078763.png)


😻

oxij commented:
Right :) Thanks!

LnL7 commented on nixos/doc/manual/default.nix:
```diff
@@ -32,8 +62,22 @@ let
+    in lib.compareLists cmp (splt a) (splt b) < 0;
+
+  # Customly sort option list for the man page.
+  optionsList = lib.sort (a: b: optionListLess a.name b.name) optionsListDesc;
```
this is the cause of the slowdown, the cost of sorting all of the options happens during every evaluation.

oxij commented:
How do I profile this in nix?

How are you sure that the problem is with the whole sort and not with
recomputation of `hasPrefix` or with `compareLists`, for instance?
Normally I would imagine that sorting shouldn't introduce too much of a
problem.

I would guess the problem is that `optionListLess` is very inefficient
when evaluated as written. Can I ask nix to partially evaluate some
code? `cmp` can be partially evaluated into a much cheaper function.

I assume this evaluation time spike influences latency of channels, so
should I make an option to disable this sorting for now?

Which documentation nixos.org displays? From a latest succeeding master,
stable, something else? I would like to keep this sorting for all
publicly visible documentation, because its awesome.

LnL7 commented:
I used this and only checked the sort, it might be not be the sort itself but some other part.
Moving this sorting to the build should work, that way it will be sorted nicely for both the manual and manpages. But adding a flag that's only enabled for the manual also works as a workaround.

```
$ nix-build pkgs/top-level/release.nix -A metrics
$ grep .lapp result/nix-support/hydra-metrics
nixos.lapp.time 1.21916 s
nixos.lapp.maxresident 455256 KiB
nixos.lapp.allocations 273427272 B
nixos.lapp.values 4435159
```

oxij commented:
> Moving this sorting to the build should work

Please elaborate, I don't understand what do you mean by that.

LnL7 commented:
Instead of sorting this in nix at evaluation time, it could be done in the docbook build or some other intermediate drv. Then the sorting only happens once and gets cached as part of the manual/manpages build.

vcunat commented:
https://nixos.org/nixos/manual/ shows the latest stable channel.

vcunat commented:
> This significantly increased evaluation time

To be more explicit, it's basically a channel blocker for nixos-unstable ATM, as Hydra almost always runs out of memory during evaluation.  Unless we can find a solution very fast, I'd temporarily revert the problematic part until we find a better way.

oxij commented:
> To be more explicit, it's basically a channel blocker for nixos-unstable ATM

Ok, this now has highest priority. Will either fix or add an option in the next 12 hours.

oxij commented:
Ok, so.

On master (ad78e52357eb3df0f457c40185504567c7da524e) with offending patches applied:

nixos.smallContainer.time 7.811 s
nixos.smallContainer.maxresident 520740 KiB
nixos.smallContainer.allocations 1008637960 B
nixos.smallContainer.values 21474162
nixos.kde.time 12.343 s
nixos.kde.maxresident 527164 KiB
nixos.kde.allocations 1176678544 B
nixos.kde.values 23892286
nixos.lapp.time 8.115 s
nixos.lapp.maxresident 521496 KiB
nixos.lapp.allocations 1019577072 B
nixos.lapp.values 21652224

With a7d75ab6489dc5834e8402db374b4d0f1e774d53 reverted (which is untrivial):

nixos.smallContainer.time 2.34 s
nixos.smallContainer.maxresident 440616 KiB
nixos.smallContainer.allocations 262088816 B
nixos.smallContainer.values 4250508
nixos.kde.time 4.52 s
nixos.kde.maxresident 525820 KiB
nixos.kde.allocations 430129400 B
nixos.kde.values 6668632
nixos.lapp.time 2.372 s
nixos.lapp.maxresident 456340 KiB
nixos.lapp.allocations 273027928 B
nixos.lapp.values 4428570

I get almost exactly the same results by replacing the comparison function with `a: b: false`.
So the problem is not with sorting, but with inefficient comparisons.

With some trivial patches of `os/doc/efficient-comparisons` branch (https://github.com/oxij/nixpkgs/commits/os/doc/efficient-comparisons) I get:

nixos.smallContainer.time 3.294 s
nixos.smallContainer.maxresident 517396 KiB
nixos.smallContainer.allocations 365512360 B
nixos.smallContainer.values 6548134
nixos.kde.time 5.367 s
nixos.kde.maxresident 530352 KiB
nixos.kde.allocations 533552944 B
nixos.kde.values 8966258
nixos.lapp.time 3.178 s
nixos.lapp.maxresident 518004 KiB
nixos.lapp.allocations 376451472 B
nixos.lapp.values 6726196

The patches are pretty trivial and the comparison function stays the same.

So, in the light of that, the question.

a7d75ab6489dc5834e8402db374b4d0f1e774d53 costs ~3.5-4x in time and mallocs. This is, indeed, too much.
`os/doc/efficient-comparisons` now costs ~1.3-1.5x in time and mallocs without losing any code quality.
Is it still too much?

Should I butcher the understendability of `optionListLess` function in favor of more optimizations?
In theory, the sky is the limit here, since `builtins.sort` doesn't influence the evaluation much.

(Also, please don't merge `os/doc/efficient-comparisons` without advance notice, its WIP.)

oxij commented:
Well, after many more experiments it doesn't get much better (-0.15s) than what `os/doc/efficient-comparisons` already had even with some advanced memoization hackery and hand-made common subexpression elimination. Random jitter thwarts all my efforts.

I can make it evaluate in 2.7s instead of 3.1s (which is statistically significant) with some minor loss of ordering (because `builins.sort` seems to be implemented in an unlucky for my hackery way), but I really prefer not to, because that unordering will get worse as more modules get added to nixos.

Lets merge `os/doc/efficient-comparisons` and then see if you still want an option to disable sorting.
Pushing sorting to docbook generation is very untrivial.

oxij commented:
PR for ofborg tests #34866.

vcunat commented:
Thanks for the quick reaction ❤️

arianvp commented:
I only get "related packages" for only 2 options in the entire manpage.     Namely `tmux` and `xen`.   Is this expected behaviour?

oxij commented:
> I only get "related packages" for only 2 options in the entire manpage. Namely `tmux` and `xen`. Is this expected behaviour?

Well, if somebody else (i.e. not me) would add more of those we would have more of them. Duh.

Meanwhile, I plan to add some more of those automatically by adding `relatedPackages` to things that are `types.package`. Having all real services have `.package` of `types.package` type seems like a good idea to me, but that requires a bunch of tiresome manual work I don't like doing.

arianvp commented:
Aah good to know. I'll see if I can sink in some boring time.

On Sun, Sep 23, 2018, 15:47 Jan Malakhovski <notifications@github.com>
wrote:

> > I only get "related packages" for only 2 options in the entire manpage.
> Namely `tmux` and `xen`. Is this expected behaviour?
>
> Well, if somebody else (i.e. not me) would add more of those we would have
> more of them. Duh.
>
> Meanwhile, I plan to add some more of those automatically by adding
> `relatedPackages` to things that are `types.package`. Having all real
> services have `.package` of `types.package` type seems like a good idea to
> me, but that requires a bunch of tiresome manual work I don't like doing.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/33898#issuecomment-423817772>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAmWo9-K-XRpv39T0dxlfDNF_4tIYBRoks5ud5DsgaJpZM4ReUY2>
> .
>
